File: js/src/serial.js
// Boundary value (de)serialization — the JS half of spec/tape-v1.md's "Value encoding".
//
// Everything crossing the boundary is encoded as JSON with revivable markers for dates;
// anything exotic degrades to an opaque marker rather than breaking the recorded call.
// The failure direction is always "the recording is a bit poorer", never "the app broke
// because it was being recorded".

const MAX_DEPTH = 16;
export const REDACTED = '[REDACTED]';

const MARKERS = new Set(['__dt__', '__date__', '__opaque__']);

function safeRepr(v, limit = 200) {
  let s;
  try {
    if (typeof v === 'function') s = `<function ${v.name || 'anonymous'}>`;
    else if (typeof v === 'symbol') s = v.toString();
    else if (typeof v === 'bigint') s = `${v}n`;
    else if (v instanceof Error) s = `${v.name}: ${v.message}`;
    else s = Object.prototype.toString.call(v);
  } catch {
    return '<unreprable>';
  }
  return s.length <= limit ? s : s.slice(0, limit - 1) + '…';
}

const opaque = (v) => ({ __opaque__: safeRepr(v) });

/**
 * Encode one boundary value.
 *
 * KNOWN, DELIBERATE LOSS: JavaScript has two nothings and the tape has one. `undefined`
 * encodes as `null`. Introducing an `__undef__` marker would be a fork in all but name —
 * the Python reader would revive it as a plain dict and every invariant would have to know
 * about it — and the alternative, dropping the key entirely, silently changes an object's
 * shape. `null` is the honest lossy choice, and it is written down here rather than
 * discovered later. If a boundary is ever found where the distinction is load-bearing,
 * adding the marker is a non-breaking change (see "Changing this" in the spec).
 */
export function toJsonable(v, depth = 0) {
  if (depth > MAX_DEPTH) return opaque(v);
  if (v === undefined || v === null) return null;

  const t = typeof v;
  if (t === 'string' || t === 'boolean') return v;
  if (t === 'number') return Number.isFinite(v) ? v : opaque(v); // NaN/±Infinity are not JSON

  if (v instanceof Date) {
    return Number.isNaN(v.getTime()) ? opaque(v) : { __dt__: v.toISOString() };
  }

  if (Array.isArray(v)) return v.map((x) => toJsonable(x, depth + 1));

  // Buffers/typed arrays: hex, tagged opaque. They are entropy or payloads, not structure.
  if (ArrayBuffer.isView(v)) {
    return { __opaque__: `<bytes ${v.byteLength}: ${Buffer.from(v.buffer, v.byteOffset, Math.min(v.byteLength, 32)).toString('hex')}>` };
  }

  if (v instanceof Map) {
    const out = {};
    for (const [k, x] of v) out[String(k)] = toJsonable(x, depth + 1);
    return out;
  }
  if (v instanceof Set) return [...v].map((x) => toJsonable(x, depth + 1));

  if (t === 'object') {
    // Only plain-ish objects are structure. A class instance is opaque: its shape is not
    // the surface a consumer reads, and reviving it faithfully is impossible anyway.
    const proto = Object.getPrototypeOf(v);
    if (proto !== null && proto !== Object.prototype) return opaque(v);

    const out = {};
    for (const [k, x] of Object.entries(v)) out[String(k)] = toJsonable(x, depth + 1);
    return out;
  }

  return opaque(v); // function, symbol, bigint
}

/** Revive a boundary value. `__opaque__` is a one-way door by design — it revives as its text. */
export function fromJsonable(v) {
  if (Array.isArray(v)) return v.map(fromJsonable);
  if (v !== null && typeof v === 'object') {
    const keys = Object.keys(v);
    if (keys.length === 1) {
      if ('__dt__' in v) return new Date(v.__dt__);
      if ('__date__' in v) return new Date(v.__date__);
      if ('__opaque__' in v) return v.__opaque__;
    }
    const out = {};
    for (const [k, x] of Object.entries(v)) out[k] = fromJsonable(x);
    return out;
  }
  return v;
}

/**
 * Apply field-name redaction rules to a jsonable tree.
 *
 * A rule that throws degrades to REDACTED: the failure direction is "masked", never
 * "leaked" and never "broke the recorded call".
 *
 * Rules MUST be idempotent. Replay re-derives the question it is about to ask, scrubs it
 * the same way, and compares against the tape — so a value that is already a mask has to
 * scrub to itself, or a redacted recording could never be replayed.
 */
export function redactJsonable(v, rules) {
  if (!rules || !Object.keys(rules).length) return v;

  if (Array.isArray(v)) return v.map((x) => redactJsonable(x, rules));

  if (v !== null && typeof v === 'object') {
    const out = {};
    for (const [k, x] of Object.entries(v)) {
      if (Object.hasOwn(rules, k)) {
        const rule = rules[k];
        if (rule === null || rule === undefined) {
          out[k] = REDACTED;
        } else {
          try {
            out[k] = rule(x);
          } catch {
            out[k] = REDACTED;
          }
        }
      } else {
        out[k] = redactJsonable(x, rules);
      }
    }
    return out;
  }

  return v;
}

/** Compact stable rendering of a chained-call argument, for `db` signatures. */
export function short(v, limit = 60) {
  let s;
  try {
    s = JSON.stringify(toJsonable(v));
  } catch {
    s = safeRepr(v);
  }
  if (s === undefined) s = 'undefined';
  return s.length <= limit ? s : s.slice(0, limit - 1) + '…';
}

export { MARKERS, MAX_DEPTH, safeRepr };

File: csharp/src/FlightRecorder/Serial.cs
// Boundary value (de)serialization — the .NET half of spec/tape-v1.md's "Value encoding".
//
// Everything crossing the recorded boundary is encoded into the JSON value model with
// revivable single-key markers for datetimes; anything exotic degrades to an opaque marker
// rather than breaking the recorded call. The failure direction is always "the recording is
// a bit poorer", never "the app broke because it was being recorded".

using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Text.RegularExpressions;

namespace FlightRecorder
{
    /// <summary>A redaction rule: a transform applied to the jsonable value, or <c>null</c> to mask.</summary>
    public delegate object? RedactTransform(object? value);

    public static class Serial
    {
        public const int MaxDepth = 16;

        /// <summary>What a masked field's value becomes under a bare (null) rule.</summary>
        public const string Redacted = "[REDACTED]";

        // Every single-key marker the value encoding uses. __undef__ exists for JavaScript,
        // which has two nothings; .NET (like Python) has one, revives it as null, and never
        // emits it. The reserved trace markers are tolerated on read, never produced here.
        private static readonly HashSet<string> Markers = new HashSet<string>
        {
            "__dt__", "__date__", "__undef__", "__opaque__",
        };

        private static readonly Regex Addr = new Regex(" at 0x[0-9A-Fa-f]+", RegexOptions.Compiled);

        // --- encode ---------------------------------------------------------------------

        /// <summary>Encode one boundary value into the tape's value model.</summary>
        public static object? ToJsonable(object? v) => ToJsonable(v, 0);

        private static object? ToJsonable(object? v, int depth)
        {
            if (depth > MaxDepth) return Opaque(v);
            switch (v)
            {
                case null:
                    return null;
                case bool b:
                    return b;
                case string s:
                    return s;
                case DateTimeOffset dto:
                    return new Dictionary<string, object?> { ["__dt__"] = Iso(dto) };
                case DateTime dt:
                    return new Dictionary<string, object?> { ["__dt__"] = IsoNaive(dt) };
            }

            var t = v.GetType();
            if (IsIntegral(t)) return Convert.ToInt64(v, CultureInfo.InvariantCulture);
            if (t == typeof(double) || t == typeof(float) || t == typeof(decimal))
            {
                var d = Convert.ToDouble(v, CultureInfo.InvariantCulture);
                return (double.IsNaN(d) || double.IsInfinity(d)) ? Opaque(v) : (object)d;
            }

            if (v is IDictionary dict)
            {
                var outMap = new Dictionary<string, object?>();
                foreach (DictionaryEntry e in dict)
                    outMap[Convert.ToString(e.Key, CultureInfo.InvariantCulture) ?? ""] = ToJsonable(e.Value, depth + 1);
                return outMap;
            }

            if (v is IEnumerable en)
            {
                var outList = new List<object?>();
                foreach (var x in en) outList.Add(ToJsonable(x, depth + 1));
                return outList;
            }

            // A plain DTO — the ordinary shape an effect returns, and an anonymous type is the
            // ordinary shape of sem/kwargs data. Record its public properties as the object a
            // consumer reads, not an opaque repr. Recurse through ToJsonable so a DateTime property
            // still gets its __dt__ marker, and camelCase the names so field-name redaction (declared
            // in the app's own lowercase vocabulary) reaches into it.
            return PocoOrOpaque(v, depth);
        }

        private static object? PocoOrOpaque(object v, int depth)
        {
            try
            {
                var props = v.GetType().GetProperties(
                    System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance);
                var outMap = new Dictionary<string, object?>();
                foreach (var p in props)
                {
                    if (!p.CanRead || p.GetIndexParameters().Length > 0) continue;
                    object? val;
                    try { val = p.GetValue(v); }
                    catch { continue; }
                    outMap[CamelCase(p.Name)] = ToJsonable(val, depth + 1);
                }
                if (outMap.Count > 0) return outMap;
            }
            catch { /* fall through to opaque */ }
            return Opaque(v);
        }

        private static string CamelCase(string name) =>
            name.Length == 0 ? name : char.ToLowerInvariant(name[0]) + name.Substring(1);

        private static bool IsIntegral(Type t) =>
            t == typeof(int) || t == typeof(long) || t == typeof(short) || t == typeof(byte)
            || t == typeof(sbyte) || t == typeof(uint) || t == typeof(ulong) || t == typeof(ushort);

        private static Dictionary<string, object?> Opaque(object? v)
        {
            var repr = ReprOrPlaceholder(v);
            repr = Addr.Replace(repr, "");
            if (repr.Length > 200) repr = repr.Substring(0, 200);
            return new Dictionary<string, object?> { ["__opaque__"] = repr };
        }

        private static string ReprOrPlaceholder(object? v)
        {
            try
            {
                return v?.ToString() ?? "null";
            }
            catch (Exception e)
            {
                return $"<unreprable {v?.GetType().Name}: {e.GetType().Name}>";
            }
        }

        // --- decode ---------------------------------------------------------------------

        /// <summary>Revive a boundary value. <c>__opaque__</c> is a one-way door by design.</summary>
        public static object? FromJsonable(object? v)
        {
            if (v is IDictionary<string, object?> map)
            {
                if (map.Count == 1)
                {
                    foreach (var kv in map)
                    {
                        switch (kv.Key)
                        {
                            case "__dt__":
                                return ParseIso(kv.Value as string);
                            case "__date__":
                                return ParseIso(kv.Value as string);
                            // JavaScript has two nothings; .NET has one. A JS tape distinguishes
                            // undefined from null; reading it here, both are simply null.
                            case "__undef__":
                                return null;
                            case "__opaque__":
                                return kv.Value;
                        }
                    }
                }
                var outMap = new Dictionary<string, object?>();
                foreach (var kv in map) outMap[kv.Key] = FromJsonable(kv.Value);
                return outMap;
            }
            if (v is IEnumerable<object?> list)
            {
                var outList = new List<object?>();
                foreach (var x in list) outList.Add(FromJsonable(x));
                return outList;
            }
            return v;
        }

        // --- redaction ------------------------------------------------------------------

        /// <summary>Apply field-name redaction rules to a jsonable tree (see <see cref="Boundary.Redact"/>).
        /// A rule that throws degrades to <see cref="Redacted"/>: the failure direction is masked,
        /// never leaked and never breaks the recorded call.</summary>
        public static object? RedactJsonable(object? v, IReadOnlyDictionary<string, RedactTransform?>? rules)
        {
            if (rules == null || rules.Count == 0) return v;

            if (v is IDictionary<string, object?> map)
            {
                var outMap = new Dictionary<string, object?>();
                foreach (var kv in map)
                {
                    if (rules.TryGetValue(kv.Key, out var rule))
                    {
                        if (rule == null)
                        {
                            outMap[kv.Key] = Redacted;
                        }
                        else
                        {
                            try { outMap[kv.Key] = rule(kv.Value); }
                            catch { outMap[kv.Key] = Redacted; }
                        }
                    }
                    else
                    {
                        outMap[kv.Key] = RedactJsonable(kv.Value, rules);
                    }
                }
                return outMap;
            }
            if (v is IEnumerable<object?> list)
            {
                var outList = new List<object?>();
                foreach (var x in list) outList.Add(RedactJsonable(x, rules));
                return outList;
            }
            return v;
        }

        /// <summary>The first forbid pattern that matches the serialized line, or null if clean.
        /// Returns the PATTERN, never the match — a tripwire that quoted the credential it caught
        /// would be the leak it exists to prevent.</summary>
        public static string? ForbiddenHit(string text, IReadOnlyList<Regex> patterns)
        {
            foreach (var p in patterns)
                if (p.IsMatch(text)) return p.ToString();
            return null;
        }

        // --- snapshots ------------------------------------------------------------------

        /// <summary>A document snapshot in the shape a well-behaved consumer reads: identity,
        /// existence, data.</summary>
        public static Dictionary<string, object?> SnapshotJsonable(string? id, bool exists, object? data) =>
            new Dictionary<string, object?>
            {
                ["id"] = id,
                ["exists"] = exists,
                ["data"] = exists ? ToJsonable(data) : null,
            };

        // --- rendering (chain signatures) -----------------------------------------------

        /// <summary>Compact stable rendering of a chain-call argument for signatures.</summary>
        public static string Short(object? v, int limit = 60)
        {
            string s;
            try { s = Json.Stringify(ToJsonable(v)); }
            catch { s = ReprOrPlaceholder(v); }
            return s.Length <= limit ? s : s.Substring(0, limit - 1) + "…";
        }

        // --- timestamps -----------------------------------------------------------------

        /// <summary>ISO-8601 with the local UTC offset — what the tape wants for aware metadata.</summary>
        public static string Iso(DateTimeOffset d) =>
            d.ToString("yyyy-MM-ddTHH:mm:ss.ffffffzzz", CultureInfo.InvariantCulture);

        /// <summary>ISO-8601 preserving the value's awareness. A DateTime with an offset renders
        /// aware; an unspecified/local one renders naive — because for an app-visible clock value
        /// the awareness is part of the value (see the `now` event in the spec).</summary>
        public static string IsoNaive(DateTime d)
        {
            if (d.Kind == DateTimeKind.Utc)
                return d.ToString("yyyy-MM-ddTHH:mm:ss.ffffff", CultureInfo.InvariantCulture) + "Z";
            return d.ToString("yyyy-MM-ddTHH:mm:ss.ffffff", CultureInfo.InvariantCulture);
        }

        internal static object ParseIso(string? s)
        {
            if (s == null) return "";
            if (DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture,
                    DateTimeStyles.RoundtripKind, out var dto)
                && (s.EndsWith("Z") || s.Contains("+") || HasNegativeOffset(s)))
                return dto;
            if (DateTime.TryParse(s, CultureInfo.InvariantCulture,
                    DateTimeStyles.RoundtripKind, out var dt))
                return dt;
            return s;
        }

        private static bool HasNegativeOffset(string s)
        {
            // A trailing -HH:MM offset (not the date's own dashes).
            var t = s.Length >= 6 ? s.Substring(s.Length - 6) : s;
            return t.Length == 6 && t[0] == '-' && t[3] == ':';
        }
    }
}
